File: 1302 - Deepest Leaves Sum/Deepest Leaves Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deepestLeavesSum(self, root: TreeNode) -> int:
        max_depth = self.maxDepth(root)
        return self.simple_traverse(root, max_depth, 1)
    
    def simple_traverse(self, root: TreeNode, max_depth: int, depth: int):
        if root is None:
            return 0
        if depth == max_depth:
            return root.val
        return self.simple_traverse(root.left, max_depth, depth + 1) + self.simple_traverse(root.right, max_depth, depth + 1)
        
    def maxDepth(self, root: TreeNode) -> int:
        if root is None:
            return 0
        return max(self.maxDepth(root.left), self.maxDepth(root.right)) + 1

File: 1302 - Deepest Leaves Sum/Deepest Leaves Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int deepestLeavesSum(TreeNode* root) {
        int max_depth = maxDepth(root);
        return simpleTraverse(root, max_depth, 1);
    }
    
    int simpleTraverse(TreeNode* root, int max_depth, int depth) {
        if (root == NULL) {
            return 0;
        }
        if (depth == max_depth) {
            return root->val;
        }
        return simpleTraverse(root->left, max_depth, depth + 1) + simpleTraverse(root->right, max_depth, depth + 1);
    }
    
    int maxDepth(TreeNode* root) {
        if (root == NULL) {
            return 0;
        }
        return max(maxDepth(root->left), maxDepth(root->right)) + 1;
    }
};
